### PR TITLE
using thread pool to avoid unneeded thread creation

### PR DESCRIPTION
**Pull request by stlankes**
Thread pools reduce the number of thread creations and the number of TLB shoot downs. Consequently, it reduces the runtime.

### CODE REVIEW

**Comment by jschwe** [comment]
I tested this commit together with https://github.com/hermitcore/libhermit-rs/pull/129/commits/3fbd6d9095a5e7b672b9785154e4f4b7c73dd547.
The results are a drastic improvement compared to https://github.com/hermitcore/libhermit-rs/issues/78#issuecomment-739274831:
| # Cores | # log lines  | # laplace time |
|:---------:|:--------:|:------------------|
| 1          |   500 | 9.8s |
| 2          | 1000 | 5.3s |
| 3          | 3450 | 4.7s |
| 4          | 5800 | 4.71s |

I guess this also indicates that the scheduler currently does not handle many threads well.

**Comment by jschwe** [comment]
Also I'd like to point out that the hyperd example added in this PR does not seem related. Did you accidentally add it to the commit?

File: examples/micro/src/benches/mem.rs [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+use std::hint::black_box;
+
+pub fn memcpy_builtin(n: usize) {
+    let v1 = vec![1u8; n];
+    let mut v2 = vec![0u8; n];
+    let src: &[u8] = black_box(&v1);
+    let dst: &mut [u8] = black_box(&mut v2);
+    
+    dst.copy_from_slice(src);
+}
```
Comment (jbreitbart) [inline examples/micro/src/benches/mem.rs:13]: You are missing a new line here.

**Review by jbreitbart — APPROVED** [review]
Fine with me, even though that is just a workaround and should not be required for the demo application.

Should we create an issue for the slow thread creation?

**Comment by stlankes** [comment]
I create an issue hermitcore/libhermit-rs#130 to solve the bad performance of the thread creation.

File: examples/micro/src/benches/mem.rs (hunk 1 above)
Comment (stlankes) [reply examples/micro/src/benches/mem.rs]: The file was located in the wrong directory. I removed it.

**Comment by stlankes** [comment]
> Also I'd like to point out that the hyperd example added in this PR does not seem related. Did you accidentally add it to the commit?

Yes, I removed the obsolete code.

**Comment by stlankes** [comment]
> I guess this also indicates that the scheduler currently does not handle many threads well.

I am not sure. I have to check it.

**Pull request merged**